### ISSUE TITLE
Use `q1_pi1` for `pi_loss`

### ISSUE THREAD

**Issue by mimoralea** [kashif/spinningup-pytorch#10]
https://github.com/kashif/firedup/blob/7011b0c8eeac5a96f995f6b44428b0334fe567a5/fireup/algos/sac/sac.py#L280

If you are trying to implement Spinning Up's version, you need to use `q1_pi` (https://github.com/openai/spinningup/blob/master/spinup/algos/sac/sac.py#L176)

```
                # Soft actor-critic losses
                pi_loss = (alpha * logp_pi - q1_pi).mean()
```

**Comment by kashif** [comment]
@mimoralea yes great observation! You are right it differs from the spinning up version and I made a PR there for it, see: openai/spinningup#85 

This was a year ago so I don't remember the reason right now and would need to go through my notes of the time, unless you can figure it out.

ps. looking forward to your book, I am sure it will be great!

**Comment by mimoralea** [comment]
Good, @kashif. I think the correct way is what you have. You can see the "Official"-Pytorch implementation uses the `min`: https://github.com/vitchyr/rlkit/blob/master/rlkit/torch/sac/sac.py#L108-L112

Same with the Official (TensorFlow) implementation: https://github.com/rail-berkeley/softlearning/blob/master/softlearning/algorithms/sac.py#L241-L244

I think opening up the pull request was the way to go, let's see what comes out of it.

Thanks for the good wishes, looking forward to the comments once it comes out.

**Comment by kashif** [comment]
Do note that the softlearning implementation came after the author's implementation which is probably what the spinningup was based on

https://github.com/haarnoja/sac/blob/master/sac/algos/sac.py#L312-L316

rail-berkeley/softlearning became the official implementation a bit later, I suspect when the updated paper came out...

**Comment by mimoralea** [comment]
Good point. I think the mix-up comes from the first version of the first paper. The second version seems to suggest two q-functions and using the min, and the second paper reinforces that.

Here are the two papers (two papers with two versions each) for future reference:

https://arxiv.org/abs/1801.01290v1:

> In the policy improvement step, we update the policy towards the exponential of the new Q-function.


https://arxiv.org/abs/1801.01290:

> Our algorithm also makes use of two Q-functions to mitigate
> positive bias in the policy improvement step that is known
> to degrade performance of value based methods (Hasselt,
> 2010; Fujimoto et al., 2018). In particular, we parameterize
> two Q-functions, with parameters θi
> , and train them independently to optimize JQ(θi). We then use the minimum of
> the Q-functions for the value gradient in Equation 6 and policy gradient in Equation 13, as proposed by Fujimoto et al.
> (2018)


https://arxiv.org/abs/1812.05905v1:

> Our algorithm makes use of two soft Q-functions to mitigate positive bias in the policy improvement
> step that is known to degrade performance of value based methods (Hasselt, 2010; Fujimoto et al.,
> 2018). In particular, we parameterize two soft Q-functions, with parameters θi
> , and train them
> independently to optimize JQ(θi). We then use the minimum of the the soft Q-functions for the
> stochastic gradient in Equation 6 and policy gradient in Equation 10, as proposed by Fujimoto
> et al. (2018). Although our algorithm can learn challenging tasks, including a 21-dimensional
> Humanoid, using just a single Q-function, we found two soft Q-functions significantly speed up
> training, especially on harder tasks

https://arxiv.org/abs/1812.05905:

> Our algorithm makes use of two soft Q-functions to mitigate positive bias in the policy improvement
> step that is known to degrade performance of value based methods (Hasselt, 2010; Fujimoto et al.,
> 2018). In particular, we parameterize two soft Q-functions, with parameters θi
> , and train them
> independently to optimize JQ(θi). We then use the minimum of the the soft Q-functions for the
> stochastic gradient in Equation 6 and policy gradient in Equation 10, as proposed by Fujimoto
> et al. (2018). Although our algorithm can learn challenging tasks, including a 21-dimensional
> Humanoid, using just a single Q-function, we found two soft Q-functions significantly speed up
> training, especially on harder tasks.
